### PR TITLE
Adding open_groups to BackendEntryPointEngine, NetCDF4BackendEntrypoint, and H5netcdfBackendEntrypoint

### LINKED ISSUE #9137: open_dict_of_datasets function to open any file containing nested groups
### Is your feature request related to a problem?

In https://github.com/pydata/xarray/issues/9077#issuecomment-2161622347 I suggested the idea of a function which could open any netCDF file with groups as a dictionary mapping group path strings to `xr.Dataset` objects.

The motivation is as follows:

- People want the new `xarray.DataTree` class to support inheriting coordinates from parent groups,
- This can only be done if the coordinates align with the variables in the child group (i.e. using `xr.align`),
- The best time to enforce this alignment is at `DataTree` construction time,
- This requirement is not enforced in the netCDF/Zarr model, so this would mean some files can no longer be opened by `open_datatree` directly, as doing so would raise an alignment error,
- _But_ we still really want users to have some way to open an arbitrary file with xarray and see what's inside (including displaying all the groups #4840).
- A simpler intermediate structure of a dictionary mapping group paths to `xarray.Dataset` objects doesn't enforce alignment, so can represent any file.
- We should add a new opening function to allow any file to be opened as this dict-of-datasets structure.
- Users can then use this to inspect "untidy" data, and make changes to the dict returned before creating an aligned `DataTree` object via `DataTree.from_dict` if they like.

### Describe the solution you'd like

Add a function like this:

```python
def open_dict_of_datasets(
    filename_or_obj: str | os.PathLike[Any] | BufferedIOBase | AbstractDataStore,
    engine: T_Engine = None,
    group: Optional[str] = None,
    **kwargs,
) -> dict[str, Dataset]:
    """
    Open and decode a file or file-like object, creating a dictionary containing one xarray Dataset for each group in the file.

    Useful when you have e.g. a netCDF file containing many groups, some of which are not alignable with their parents and so the file cannot be opened directly with ``open_datatree``.

    It is encouraged to use this function to inspect your data, then make the necessary changes to make the structure coercible to a `DataTree` object before calling `DataTree.from_dict()` and proceeding with your analysis.

    Parameters
    ----------
    filename_or_obj : str, Path, file-like, or DataStore
        Strings and Path objects are interpreted as a path to a netCDF file or Zarr store.
    engine : str, optional
        Xarray backend engine to use. Valid options include `{"netcdf4", "h5netcdf", "zarr"}`.
    group : str, optional
        Group to use as the root group to start reading from. Groups above this root group will not be included in the output.
    **kwargs : dict
        Additional keyword arguments passed to :py:func:`~xarray.open_dataset` for each group.

    Returns
    -------
    dict[str, xarray.Dataset]

    See Also
    --------
    open_datatree()
    DataTree.from_dict()
    """
    ...
```

This would live inside `backends.api.py`, and be exposed publicly as a top-level function along with the rest of `open_datatree`/`DataTree` etc. as part of #9033.

The actual implementation could re-use the code for opening many groups of the same file performantly from #9014. Indeed we could add a `open_dict_of_datasets` method to the `BackendEntryPoint` class, which uses pretty much the same code as the existing `open_datatree` method added in #9014 but just doesn't actually create a `DataTree` object.

### Describe alternatives you've considered

Really the main alternative to this is not to have coordinate inheritance in `DataTree` at all (see [9077](https://github.com/pydata/xarray/issues/9077)), in which case `open_datatree` would be sufficient to open any file.

---

The name of the function is up for debate. I prefer nothing with the word "datatree" in it since this doesn't actually create a `DataTree` object at any point. (In fact we could and perhaps should have implemented this function years ago, even without the new `DataTree` class.) The reason for not calling it "`open_as_dict_of_datasets`" is that we don't use "as" in the existing `open_dataset`/`open_dataarray` etc.

### Additional context

cc @eni-awowale @flamingbear @owenlittlejohns @keewis @shoyer @autydp

### PR DESCRIPTION
**Pull request by eni-awowale**
- [x] Closes #9137 and in support of #8572
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

@TomNicholas, @shoyer, @owenlittlejohns, and @flamingbear

### CODE REVIEW

File: xarray/backends/h5netcdf_.py
Comment (TomNicholas) [inline xarray/backends/h5netcdf_.py:463]: This `groups_dict` should be created by calling `self.open_groups`, to avoid code duplication.

File: xarray/backends/api.py
Comment (flamingbear) [inline xarray/backends/api.py:1726]: I'd move this up above open_mfdataset

**Comment by eni-awowale** [comment]
I know we touched on this a bit during our meeting on Tuesday but I think tests for `open_groups` in `netCDF_.py` and `h5netcdf.py` should probably go in `test_backends_datatree.py`. We mentioned we could create separate test files for it but it looks like `test_backends_datatree.py`, has an empty class for each of these backends and it looks like they would fit nicely in here.

[`test_backends_datatree.py`](https://github.com/pydata/xarray/blob/main/xarray/tests/test_backends_datatree.py#L66-73)
```
@requires_netCDF4
class TestNetCDF4DatatreeIO(DatatreeIOBase):
    engine: T_DataTreeNetcdfEngine | None = "netcdf4"


@requires_h5netcdf
class TestH5NetCDFDatatreeIO(DatatreeIOBase):
    engine: T_DataTreeNetcdfEngine | None = "h5netcdf"
```

Let me know if I have that right and what you all think.
@TomNicholas, @shoyer, @keewis, @flamingbear, @owenlittlejohns

**Comment by TomNicholas** [comment]
Sounds good @eni-awowale 

You may even be able to use basically the same groups test for multiple backends by adding it to `DataTreeIOBase`...

> If it's okay with everyone I think we should also include a small netCDF file in `test/data` that has groups. 

Having an included file to test against is a good idea, we just want it to be small.

> `example_1.nc` and `bears.nc` don't have groups.

Where are these currently? We don't want to add files to the main repo - preferring to put them here https://github.com/pydata/xarray-data

**Comment by eni-awowale** [comment]
> You may even be able to use basically the same groups test for multiple backends by adding it to DataTreeIOBase...

So, instead of adding the same type of tests to `TestNetCDF4DatatreeIO` and `TestH5NetCDFDatatreeIO` I can add them directly to `DataTreeIOBase`, if the are not engine specific? Okay, that makes sense!

@TomNicholas `example_1.nc` and `bears.nc` are here [test/data](https://github.com/pydata/xarray/tree/main/xarray/tests/data).

> We don't want to add files to the main repo - preferring to put them here

Okay, great there might already be some some sample files in there with groups that I can use!

**Comment by TomNicholas** [comment]
> @TomNicholas example_1.nc and bears.nc are here [test/data](https://github.com/pydata/xarray/tree/main/xarray/tests/data).

I didn't even know those existed... @max-sixty I know you have thought about example datasets for testing purposes - do you have an opinion on whether new files for testing should go in that directory or the separate repository?

**Comment by shoyer** [comment]
> > @TomNicholas example_1.nc and bears.nc are here [test/data](https://github.com/pydata/xarray/tree/main/xarray/tests/data).
> 
> I didn't even know those existed... @max-sixty I know you have thought about example datasets for testing purposes - do you have an opinion on whether new files for testing should go in that directory or the separate repository?

We want tests to be runable without a network connection. If we need new files for testing, which I agree would be a good idea in this case, please add them here. Just keep them as small as possible (the current test files are all under 10 KB in size, most under 1 KB).

**Comment by shoyer** [comment]
> Where are these currently? We don't want to add files to the main repo - preferring to put them here https://github.com/pydata/xarray-data

xarray-data is for sample/tutorial datasets, which can be much larger (enough data to make an interesting plot) than what we use for test data.

File: xarray/backends/h5netcdf_.py
Comment (eni-awowale) [inline xarray/backends/h5netcdf_.py:506]: Added this since we need to read the root group once and add it to the dictionary.

File: xarray/backends/h5netcdf_.py
Comment (eni-awowale) [inline xarray/backends/h5netcdf_.py:470]: Merged these backend specific keyword arguments from #9199

File: xarray/backends/plugins.py
Comment (eni-awowale) [inline xarray/backends/plugins.py:206]: Made this a little more descriptive because I kept getting confused lol

File: xarray/tests/test_backends_datatree.py
Comment (eni-awowale) [inline xarray/tests/test_backends_datatree.py:66]: I could always add more tests for `open_datatree` but I think @shoyer really covered everything in `test_datatree.py`

File: xarray/tests/test_backends_datatree.py
Comment (eni-awowale) [inline xarray/tests/test_backends_datatree.py:85]: I added `"test_data_not_aligned.nc"` to the `tests/data` directory. I could definitely add more test files and tests but didn't want to bloat the directory with too many test data files.

File: xarray/backends/h5netcdf_.py
Comment (eni-awowale) [reply xarray/backends/h5netcdf_.py:506]: Thanks! How would you suggest prepending the root path to `_inter_nc_groups`?

Comment (TomNicholas) [reply xarray/backends/h5netcdf_.py:507]: You can use the `NodePath` object and pathlib-like syntax, e.g. `<first-half> / <second-half>`

https://github.com/pydata/xarray/blob/8c8d097816a70e35ef60de301503aa33f662857c/xarray/core/treenode.py#L26

File: xarray/backends/plugins.py
Comment (TomNicholas) [reply xarray/backends/plugins.py:206]: Good idea, but "download engine" isn't quite accurate in my opinion. How about

`f"unrecognized backend engine {engine} must be one of your installed backend engines:`

File: xarray/backends/common.py
Comment (TomNicholas) [inline xarray/backends/common.py:538]:
```suggestion
    def open_groups_as_dict(
```
This was @shoyer's suggestion - keep the top-level public-facing function called `open_groups`, but change the name of this method to help avoid confusion with the `open_group` method defined on the zarr backend.

File: xarray/backends/common.py
Comment (TomNicholas) [inline xarray/backends/common.py:545]: Some description of the motivation + relationship to `open_datatree` for developers would be good here, e.g.

```suggestion
        """
        Opens a dictionary mapping from group names to Datasets.
        
        Called by :py:func:`~xarray.open_groups`. This function exists so as to provide a universal way to open all groups in a file, before applying any additional consistency checks or requirements necessary to create a `DataTree` object (typically done using :py:meth:`~xarray.DataTree.from_dict`). 
        """
```

File: xarray/backends/h5netcdf_.py
Comment (TomNicholas) [inline xarray/backends/h5netcdf_.py:476]: These shouldn't be here right? They should all fall under `**kwargs``

File: xarray/backends/h5netcdf_.py
Comment (TomNicholas) [inline xarray/backends/h5netcdf_.py:499]: @eni-awowale this is how you should join paths

File: xarray/backends/h5netcdf_.py
Comment (TomNicholas) [reply xarray/backends/h5netcdf_.py:476]: Or maybe they should be in the specific backend but not in `common.py`?

File: xarray/backends/api.py
Comment (TomNicholas) [inline xarray/backends/api.py:837]: We _could_ have a default implementation here that calls `open_groups`, i.e.

```suggestion
    groups_dict = backend.open_datatree(filename_or_obj, **kwargs)
    return DataTree.from_dict(groups_dict)
```

The idea being that then backend developers don't actually have to implement `open_datatree` if they have implemented `open_groups`...

This was sort of discussed here (@keewis) https://github.com/pydata/xarray/pull/7437#discussion_r1279083937, but this seems like an rabbit hole that should be left for a future PR.

Comment (keewis) [reply xarray/backends/api.py:837]: not really, I was arguing that given any one of `open_dataarray`, `open_dataset` and `open_datatree` allows us to provide (somewhat inefficient) default implementations for the others. However, `open_groups` has a much closer relationship to `open_datatree`, so I think having a default implementation for `open_datatree` is fine (we just need to make sure that a backend that provides neither `open_groups` nor `open_datatree` doesn't complain about `open_groups` not existing if you called `open_datatree`).

So yeah, this might become a rabbit hole.

Comment (TomNicholas) [reply xarray/backends/api.py:837]: Okay I see. That seems related, but also like a totally optional convenience feature that we should defer to later.

File: xarray/backends/h5netcdf_.py
Comment (keewis) [reply xarray/backends/h5netcdf_.py:507]: we can use [`prepend` from the `itertools` recipes](https://docs.python.org/3/library/itertools.html#itertools-recipes)

Comment (TomNicholas) [reply xarray/backends/h5netcdf_.py:507]: > we can use [prepend from the itertools recipes](https://docs.python.org/3/library/itertools.html#itertools-recipes)

We could, that's a neat solution for a general iterable, but given that we know we're dealing with paths here wouldn't it be better to use the dedicated class we have for handling node paths? (This is not particularly important though)

Comment (keewis) [reply xarray/backends/h5netcdf_.py:507]: no, I meant that `_iter_nc_groups` yields nodes / node paths, and we can use `prepend` to prepend the root node's path (as a `NodePath`)

File: xarray/backends/h5netcdf_.py
Comment (eni-awowale) [reply xarray/backends/h5netcdf_.py:476]: Yeah, so these were added from PR #9199 for adding back the backend specific keyword arguments. I pulled this into my branch after it was merged to main. But they are not in `common.py` , they are consolidated as `**kwargs` in `common.py`.

File: xarray/backends/h5netcdf_.py
Comment (eni-awowale) [reply xarray/backends/h5netcdf_.py:507]: I don't think I am completely following this thread. But what if instead of reopening the file_obj we open the store like: 
```suggestion
        ds = open_dataset(store, **kwargs)
        groups_dict = {str(parent): ds}
```

Comment (keewis) [reply xarray/backends/h5netcdf_.py:508]: that would work, except now we have a separate code path for the parent group compared to the child groups (and `open_dataset` applies the decoders if I'm not wrong?). Looking at this again, I now believe sticking `yield parent` in between these two lines could fix this in a much simpler way: https://github.com/pydata/xarray/blob/c508cc6a2e3000a9d87d2f8c611aae8733be07bf/xarray/backends/common.py#L134-L135

**Comment by eni-awowale** [comment]
Hi folks! I think this PR is a couple commits away from merging but they are a couple things to sort out. 

### `mypy` checks failing
The `open_groups` function returns a `dict[str, Dataset]` type. However, the return type for `DataTree.from_dict()` is a `MutableMapping[str, Dataset | DataArray | DataTree | None]` and we call `DataTree.from_dict()` in our `open_datatree()` backend functions, so `mypy` flags this as having an incompatible type. See below.
`"from_dict" of "DataTree" has incompatible type "dict[str, Dataset]"; expected "MutableMapping[str, Dataset | DataArray | DataTree[Any] | None]"  [arg-type]`
I am not sure exactly how to proceed with this since updating the return type of `DataTree.from_dict()` causes other mypy checks to fail because the are expecting the `MutuableMapping` return type.

### Some checks are failing but seem unrelated to the PR changes?
I noticed there are tests failing, but these seem to be consistent with the checks failing in main and are related to `test_computations.py`. The other test I see failing are the ubuntu 3.12 flaky tests. Are these failures we can ignore?

### Updating `_iter_nc_groups`
I added the suggestion from @keewis  so `_iter_nc_groups` is returning the parent node and now we don't have to reopen the store twice 🎉

cc
@TomNicholas, @keewis, @shoyer, @flamingbear

**Comment by keewis** [comment]
> Are these failures we can ignore?

`flaky` is something that may fail at random, mostly because of server or network issues. Usually it will pass again on the next run (I reran it for you, let's see what happens). The other test failures are indeed also on `main`, so we can ignore it here.

Edit: `flaky` passes now, so I believe you can ignore the remaining test failures

> I am not sure exactly how to proceed with this since updating the return type of `DataTree.from_dict()` causes other `mypy` checks to fail because the are expecting the `MutableMapping` return type.

Maybe the typing crowd can help? cc @max-sixty, @Illviljan, @headtr1ck

**Comment by Illviljan** [comment]
> ### `mypy` checks failing
> 
> The `open_groups` function returns a `dict[str, Dataset]` type. However, the return type for `DataTree.from_dict()` is a `MutableMapping[str, Dataset | DataArray | DataTree | None]` and we call `DataTree.from_dict()` in our `open_datatree()` backend functions, so `mypy` flags this as having an incompatible type. See below. `"from_dict" of "DataTree" has incompatible type "dict[str, Dataset]"; expected "MutableMapping[str, Dataset | DataArray | DataTree[Any] | None]" [arg-type]` I am not sure exactly how to proceed with this since updating the return type of `DataTree.from_dict()` causes other mypy checks to fail because the are expecting the `MutuableMapping` return type.

It's this type of problem: https://stackoverflow.com/questions/73603289/why-doesnt-parameter-type-dictstr-unionstr-int-accept-value-of-type-di
`from_dict` should probably switch from `MutableMapping` to `Mapping` instead.

File: xarray/tests/test_backends_datatree.py
Comment (Illviljan) [inline xarray/tests/test_backends_datatree.py:76]:
```suggestion
    def test_open_datatree(self, tmpdir) -> None:
```

File: xarray/tests/test_backends_datatree.py
Comment (Illviljan) [inline xarray/tests/test_backends_datatree.py:119]:
```suggestion
    def test_open_groups(self, tmpdir) -> None:
```

File: xarray/tests/test_backends_datatree.py
Comment (Illviljan) [inline xarray/tests/test_backends_datatree.py:179]:
```suggestion
    def test_open_groups_to_dict(self, tmpdir) -> None:
```

File: xarray/core/datatree.py
Comment (Illviljan) [inline xarray/core/datatree.py:1100]: Mypy is correct here, a Mapping does not include a `.pop` and ignoring the typing errors doesn't solve the bug. 
xarray uses Mappings frequently, for example `xr.core.utils.FrozenDict(dict(a=3)).pop("a", None)` so it's a real issue.

Either you explicitly convert it to dict i.e. `d_cast = dict(d)` or refactor the code to not use the `.pop` since I'm not so sure it's needed.

Comment (eni-awowale) [reply xarray/core/datatree.py]: Thanks, I updated it to explicitly covert the type to a dict. The mypy3.9 tests are still passing but the CI Mypy check seems to be returning the same error as before explicitly converting it to a dict.

**Comment by eni-awowale** [comment]
@TomNicholas  and @keewis this issue we just moved over #9336 seems to be related to the latest batch of test failures.

**Comment by eni-awowale** [comment]
Yay all checks are passing! Does someone want to give this a quick look before merging?
cc
@TomNicholas, @keewis, @owenlittlejohns, @shoyer and @flamingbear

File: xarray/backends/common.py
Comment (headtr1ck) [inline xarray/backends/common.py:542]: Maybe we should not make the same mistake as with `open_dataset` and prevent liskov errors.

```suggestion
```

If the abstract method supports any kwargs, so must all subclass implementations, which is not what we want.

File: xarray/backends/h5netcdf_.py
Comment (headtr1ck) [inline xarray/backends/h5netcdf_.py:477]: This should be obsolete as well, when you remove it from the abstract method.

File: xarray/backends/netCDF4_.py
Comment (headtr1ck) [inline xarray/backends/netCDF4_.py:716]: Same here.

**Review by headtr1ck — APPROVED** [review]

File: xarray/backends/common.py
Comment (eni-awowale) [reply xarray/backends/common.py:542]: @headtr1ck from my understanding I think the `**kwargs` were added back to fix this issue #9135

Comment (headtr1ck) [reply xarray/backends/common.py:542]: Hmmm, Not sure.
But since this is the same problem in all other backend methods, I'm fine with leaving it as it is (and possibly change it in a future PR all together).

Comment (eni-awowale) [reply xarray/backends/common.py:542]: Sounds good we can revisit this on another PR.

**Pull request merged**
